### ISSUE TITLE
Lazily iterate over tests instead of collecting

### ISSUE THREAD

**Issue by gakonst** [foundry-rs/foundry#141]
Instead of collecting [here](https://github.com/gakonst/foundry/blob/master/forge/src/runner.rs#L196-L211) and [here](https://github.com/gakonst/foundry/blob/0785187319a9cdbc070ad40040a360b2e23ddaf0/forge/src/multi_runner.rs#L151-L157), it’d be nice if we could have an iterator which we can stream over

That'd allow us to then iterate in parallel [here](https://github.com/gakonst/foundry/blob/master/cli/src/cmd/test.rs#L189) for nicer UX & speed

**Comment by tansawit** [comment]
Is this still applicable? The codebase seemed to have changed a lot since the issue was open.

**Comment by gakonst** [comment]
I've updated the links, thanks for flagging.

**Comment by lattejed** [comment]
If no one is on this, I'll pick it up

**Comment by CalebEverett** [comment]
I was taking a look at this as well and was wondering what the right parallelization strategy was with regard to the `evm`. Does it make sense to try to just use a single one or will that limit the potential speedup and would it be better to have multiple, potentially one per `ContractRunner` or even multiple within a `ContractRunner` since the state gets reset before each test runs.

**Comment by lattejed** [comment]
@CalebEverett I need to look at it in depth before I can comment. Did you want to pick this one up? I don't mind either way.

**Comment by CalebEverett** [comment]
@lattejed I'd be up for trying to figure it out, but it will probably take me longer to get all the way there than it would you. If you take a look at it and think it is something quick for you, in the interest of helping the repo, by all means take it. I'd be curious on the solution. But otherwise, happy to keep working on it.

**Comment by lattejed** [comment]
@CalebEverett cool, I'll take a look. Worth discussing here regardless.

**Comment by gakonst** [comment]
I think we can start by naively creating a factory function `new_evm(<..args>) -> impl Evm` (I think Clone'ing the EVM might not be an option) which gets passed in the runner/multi_runner.rs and gets called in each test. 

We are already clone'ing and passing the init state for each test, so I think we have most of the state related components set up.

Highest performance gain will be by parallelizing the execution of the fuzz tests.

**Comment by lattejed** [comment]
Good suggestions, thanks. Looking at it now Evm is stateful / reset inside the runner for each test. Based on your comments I'll assume there are no issues instantiating a number of Evms in parallel so it should be straightforward to make these changes.

**Comment by gakonst** [comment]
Yes exactly, I think that because of the resetting of the state being already done it should be doable

**Comment by lattejed** [comment]
Currently looking at implementing `Clone` for `Evm` as it may be cleaner overall. 

Main challenge with adding parallelism here will be managing lifetimes. `Clone` may be more straightforward than working with passing around state to init new Evms, but I need to experiment more to see what works best.

**Comment by brockelmore** [comment]
@lattejed I would recommend just instantiating *inside* the test. You probably will run into `Send`+ `Sync` issues otherwise as I tried to hack something together yesterday using clone to no avail.

**Comment by CalebEverett** [comment]
That was the approach I was working on as well as a starting place, removing the passing on an instantiated vm to [ContractRunner](https://github.com/gakonst/foundry/blob/cb279a21b562df7f57b5880584c722d77b93018c/forge/src/runner.rs#L139) and just creating one inside [run_test](https://github.com/gakonst/foundry/blob/cb279a21b562df7f57b5880584c722d77b93018c/forge/src/runner.rs#L231) with a `new_evm` method implemented on `ContractRunner`, but hadn't sorted out the trait bounds yet.

I was starting with just the sputnik once since it didn't require any additional arguments to create.

```rust
impl<'a, S: Clone, E: Evm<S>> ContractRunner<'a, S, E> {
    pub fn new_evm() -> impl Evm<S> {
        vm()
    }
```

but that is saying:
```
the trait bound `Executor<MemoryStackStateOwned<'_, CheatcodeBackend<MemoryBackend<'_>>>, CheatcodeHandler<StackExecutor<'_, '_, MemoryStackStateOwned<'_, CheatcodeBackend<MemoryBackend<'_>>>, BTreeMap<H160, for<'r, 's> fn(&'r [u8], std::option::Option<u64>, &'s evm_adapters::sputnik::evm::Context, bool) -> Result<PrecompileOutput, PrecompileFailure>>>>>: evm_adapters::Evm<S>` is not satisfied
```

which was weird since that is the same way the the evm that get's passed into `ContractRunner` gets created and the trait bound is satisfied. I'm probably missing something basic in satisfying the trait bound in the new method.

**Comment by gakonst** [comment]
If you push the code we can maybe take a look, not sure what this could be either. Maybe a missing `Sync + Send` trait bound?

**Comment by CalebEverett** [comment]
[Here](https://github.com/CalebEverett/foundry/blob/2f880a17a7595ef0e7c5e6b39f6b4d283ccfcf04/forge/src/runner.rs#L171) is the new method. Starting with just keeping everything as is and getting the [creation of the vm](https://github.com/CalebEverett/foundry/blob/2f880a17a7595ef0e7c5e6b39f6b4d283ccfcf04/forge/src/runner.rs#L202) and passed into [run_test](https://github.com/CalebEverett/foundry/blob/2f880a17a7595ef0e7c5e6b39f6b4d283ccfcf04/forge/src/runner.rs#L236) going.

**Comment by brockelmore** [comment]
the primary issue with the above is that we lose the command line configurations (like forking).

This is semi-ok but for forking we likely want a single evm anyway because otherwise caching becomes impossible in its current form (there is another fork caching implementation floating around some place that isnt used but may solve this)

**Comment by CalebEverett** [comment]
I was thinking `new_evm` would accept arguments. I was messing around with putting the evm in a mutex and sharing it between threads, but If there is only one evm, how much speed up is there to be had over the current implementation? I was working on naively testing what the tradeoff was of having to create evms versus the benefit of having the tests themselves run in parallel as a first pass.

If I understand the [forking requirement](https://github.com/gakonst/foundry/blob/e1fb71dd622c407a92046b85fe2e2f2e5989eb82/evm-adapters/README.md#sputniks-cached-forking-backend), could you still make that work by storing that as the same initial `state`? `new_evm` would use whatever is in `state` as its initial state and if it received a forking argument, it would just grab the forked state once and then use that on each creation.

**Comment by gakonst** [comment]
I'd use the "SharedBackend" that @mattsse wrote, it's a cloneable forking backend which caches and deduplicatew requests and responses. The explicit reason it was created was for when we have multiple tests running in parallel to optimize network traffic.

**Comment by lattejed** [comment]
The main issue I had with passing in a fn was that it looked like there were multiple factory functions per vm in the different implementations -- it looked like it might be necessary to pass around additional config for `fn() -> Evm` to work. Not that big of a deal but I was trying to avoid it. Maybe this won't be an issue in practice however., i.e., each vm has a single config for testing.

So I played around with passing in a closure so vm setup could be kept in the same places it currently lives. There are lifetime issues with that since closures are short lived out of the box. This is probably not worth pursuing.

@brockelmore you're likely right. I stubbed out `Clone` but didn't try to pass it to a new thread yet. I like the ergonomics of it a lot better (clone, reset / copy state) but it's likely to be more trouble than it's worth, assuming it's possible at all.

I'm currently wrapping up a late holiday and have a long drive today. It'll be a bit before I can do a deep dive and make any meaningful recommendations. @CalebEverett looks like you're making good progress, I'll see where you're at / what I can help with when I'm back online.

**Comment by lattejed** [comment]
Spending some more time with it now. Some thoughts:

Instead of configuring an `Evm` in the test `Cmd` and then having it passed around belonging to `[Multi]ContractRunner` I think it makes sense to pass around a new `EvmConfig` struct (whatever it should be called) and make `ContractRunner` able to init new Evms as needed, including inside new threads. Maybe `new_evm` needs to be a free function to simplify this. This would partially defeat the purpose of passing around `trait Evm` but should simplify things.

@gakonst I'm not sure where you were suggesting adding `new_evm(<..args>) -> impl Evm` so maybe we're talking about the same thing.

The downside I guess is making the runners and Evms more tightly coupled. But the less state that needs to be passed between threads the better.

I don't think this magically fixes everything, there's a lot of complex state here that has to be passed to new threads and we know that Rust is great partially because it's very restrictive about these things.

Thoughts are more than welcome. I'm trying to look for a least intrusive way to add parallelism, though I don't currently have in mind any 'messy but works' solutions either.

**Comment by lattejed** [comment]
Deleted a comment. Need to play around more.

**Comment by lattejed** [comment]
Ok, a bit clearer now.

Afaik `fn() -> impl Evm` can only be used to return homogenous types. The common way to return heterogenous types is to use `fn() -> Box<dyn some_trait>` (or an `Enum` which won't really work here).

However, in this case we need to return a `Box<evm_adapters::Evm<State>>` and `evm_adapters::Evm<State>` is considered 'object unsafe' or unboxable due the use of generics and some other issues.

So, passing around a factory function that can return `Box<evm_adapters::Evm<State>>` isn't going to work without modifying the trait. Rust forbids using generics with dynamic dispatch, so this is likely a no-go.

An alternative would be to move vm creation into the test runners where necessary. That somewhat defeats the purpose of having the vm's conform to `evm_adapters::Evm<State>`. However, it's still not an issue passing these trait objects to functions, so there's still gain there.

If vms have to be instantiated in multiple places, a macro could clean things up. I'm still on the fence about how awful that is.

An alternative would be to try to get everything implementing `Evm` to be `Clone`. I'm think we can work around `Send+Sync` issues by cloning and moving. However, `Clone` is probably nearly as problematic as `Send+Sync` to conform to.

I also considered some kind of wrapper / builder,  but I believe the basic issues will remain. 

I'm reluctant to tightly couple the runners and evm creation, but maybe that's the only way out of here w/o a lot of rewrite. 

Any / all input welcome here. Also happy to be told I've overlooked something obvious.

**Comment by lattejed** [comment]
I'm going to call `Clone` a dead end -- `derive` requires upstream changes and would be difficult to implement manually due to the use of generics.

Currently working on moving evm creation into the test runners. Probably going to go the route of adding a macro so evm creation logic doesn't have to be replicated.

**Comment by gakonst** [comment]
Ideally we'd avoid the macro, can you share maybe what the roadblock with moving the evm creation function (instead of doing a macro) is?

**Comment by lattejed** [comment]
I flushed my changes already, but you can drop sth like this into `cli/src/cmd/test.rs`:

```
pub fn new_evm<S: Clone, R>(opts: BuildArgs, evm_opts: EvmOpts) -> Box<dyn evm_adapters::Evm<S, ReturnReason = R>> {
    let mut cfg = utils::sputnik_cfg(opts.compiler.evm_version);
    let vicinity = evm_opts.vicinity().unwrap();
    Box::new(utils::sputnik_helpers::evm(&evm_opts, &mut cfg, &vicinity).unwrap())
        as Box<dyn evm_adapters::Evm<S, ReturnReason = R>>
}
```

`Box<dyn trait>` is required b/c of how Rust deals with heterogenous return types. The simpler `-> impl Evm` will only work with homogenous types. 

You could create multiple factory functions, but will still need some config passed along with it and some logic in place.

**Comment by lattejed** [comment]
To be clear, a macro isn't required. I was just considering it if evm creation logic has to be duplicated in multiple places. 

The 'have to' afaict here is to move the creation logic into the runners.

**Comment by mattsse** [comment]
yeah the Evm trait is not object-safe

since we only use the Evm trait, would it be possible to do something like

```rust

enum Evm {Sputnik(..), Evmodin(..)}

```

I guess that would require wrapper types for the state as well, so maybe too complex

Since they are both mutually exclusive, can make this work with cfg type aliases

```rust
#[cfg(feature = "sputnik")
type Evm = Sputnik<..>;

#[cfg(feature = "evmodin")
type Evm = Evmodin<..>;
```
same for factories

**Comment by lattejed** [comment]
Yeah, I was wondering about that. This

```
match evm_opts.evm_type {
    #[cfg(feature = "sputnik-evm")]
    EvmType::Sputnik => {}
    #[cfg(feature = "evmodin-evm")]
    EvmType::EvmOdin => {}
}
```
made me think they weren't mutually exclusive, but if they are that gives some more leeway.

I'll see if the aliases you're suggesting can work. Thanks for the input @mattsse

**Comment by CalebEverett** [comment]
@lattejed I managed to get a super crude version of this running with the tests parallelized in `run_tests` and sputnik vms created in `run_test` and `run_fuzz_test`. I was working on getting the initial state with the deployed contracts passed into `run_tests` so it can be cloned for each `run_test` call and was then going to circle back to getting the cli args passed in and generalizing to include evmodin.

**Comment by gakonst** [comment]
@CalebEverett pushing this code would be very valuable. I'm very curious to hear how long [solmate](https://github.com/Rari-Capital/solmate) tests take with your code

**Comment by CalebEverett** [comment]
@gakonst [Here](https://github.com/CalebEverett/foundry/commit/9c66cea8d29d8d8b91b55c4010b16477dd5e0c9c) is what I've got - super crude, intended as concept demo only, to be refined, etc. I think the state with the deployed contracts and cli args need to get completed before being able to try out in the wild.

**Comment by brockelmore** [comment]
after some serious hacking on caleb's impl, i was able to run solmate is 10.8 seconds, vs 93secs on master @gakonst

**Comment by CalebEverett** [comment]
@brockelmore - sorry about that - would love to see where you got to and try to get it cleaned up.

**Comment by lattejed** [comment]
@CalebEverett awesome, I'll take a look at what you've done.

**Comment by brockelmore** [comment]
https://github.com/gakonst/foundry/tree/brock/par_evm

take a look at my branch, should just work

**Comment by lattejed** [comment]
@brockelmore Thanks, really appreciate it. Was getting grindy, mostly due to lack of familiarity with the project.

Are you going to merge this or are you suggesting I pull your changes into a PR. I can address your note there, sure.

Looks like you pulled contract deployment out of `MultiContractRunnerBuilder` and I don't see it replaced anywhere. Was that optional?

**Comment by brockelmore** [comment]
@lattejed feel free to just open PRs against my PR. the deployment occurs in `new_sputnik_evm`, which is where we should grab the init logs. You cant do it in multicontractrunner because the address for the test contract will be wrong as it would deploy all of them sequentially with different nonces relative to a fresh evm

**Comment by lattejed** [comment]
@brockelmore FYI over a very basic set of tests I'm seeing a ~2.5x speedup 👍 

I don't see a PR related to this from you. I'm currently working off your branch and will put together sth from there.

**Comment by lattejed** [comment]
@brockelmore this is ready to merge into your branch. Fixed a couple of minor issues and updated the tests.

**Comment by gakonst** [comment]
no longer relevant since the tests parallelization PR
